app/routines/confirm_by_pin.rb

```diff
@@ -0,0 +1,41 @@
+class ConfirmByPin
+  lev_routine
+
+  uses_routine ConfirmContactInfo
+
+  MAX_PIN_FAILURES = 5
+
+  def self.sequential_failure_for(contact_info)
+    SequentialFailure.confirm_by_pin
+                     .find_or_initialize_by_reference(contact_info.value).tap do |sf|
+      sf.num_failures_allowed = MAX_PIN_FAILURES
+    end
+  end
+
+  protected
+
+  def exec(contact_info:, pin:)
+    return if contact_info.confirmed?
+
+    sequential_failure = self.class.sequential_failure_for(contact_info)
+
+    if !sequential_failure.attempts_remaining?
+      fatal_error(code: :no_pin_confirmation_attempts_remaining)
+    else
+      if contact_info.confirmation_pin == pin
+        after_transaction do
+          sequential_failure.reset!
+        end
+
+        run(ConfirmContactInfo, contact_info)
+      else
+        after_transaction do
+          sequential_failure.increment!
+        end
+
```

Comment:
The `sequential_failure.reset!` call should probably be inside the routine's transaction, but the `increment!` call needs to be afterwards.  The reason is that the `fatal_error` call will rollback the routine transaction which will undo the increment (which we don't want).